### PR TITLE
Mongodb subscriber

### PR DESCRIPTION
**Pull request by achandras**

### CODE REVIEW

**Comment by mpetazzoni** [comment]
This LGTM. @keitwb if you agree please squash and merge.

File: lib/mongodb/instrumentation/command_subscriber.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        @tracer = tracer
+
+        @requests = {}
+      end
+
+      def started(event)
+        # start command span
+        tags = {
+          # opentracing tags
+          'component' => 'mongodb-instrumentation',
```
Comment (mpetazzoni) [inline lib/mongodb/instrumentation/command_subscriber.rb]: This should just be `mongodb`, or `ruby-mongodb` I think

File: lib/mongodb/instrumentation/command_subscriber.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          'db.instance' => event.database_name,
+          'db.statement' => event.command,
+          'db.type' => 'mongo',
+          'span.kind' => 'client',
+
+          # extra info
+          'command_name' => event.command_name,
```
Comment (mpetazzoni) [inline lib/mongodb/instrumentation/command_subscriber.rb]: You might want to namespace those additional tags:

```
mongo.command.name
mongo.operation.id
mongo.request.id
```

File: lib/mongodb/instrumentation/command_subscriber.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        @requests[event.request_id] = span
+      end
+
+      def succeeded(event)
+        return if @requests[event.request_id].nil?
+
+        # tag the reported duration, in case it differs from what we saw
+        # through the notifications times
+        span = @requests[event.request_id]
+        span.set_tag("duration", event.duration)
```
Comment (mpetazzoni) [inline lib/mongodb/instrumentation/command_subscriber.rb]: I'd recommend calling this `took`, to differentiate from the span's duration. Also, what is the unit here? If it's milliseconds for example, call `took.ms`.

File: lib/mongodb/instrumentation/command_subscriber.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        span.finish()
+        @requests.delete(event.request_id)
+      end
+
+      def failed(event)
+        return if @requests[event.request_id].nil?
+
+        # tag the reported duration and any error message that came through
+        span = @requests[event.request_id]
+        span.set_tag("duration", event.duration)
```
Comment (mpetazzoni) [inline lib/mongodb/instrumentation/command_subscriber.rb]: Same comment about duration

File: lib/mongodb/instrumentation/command_subscriber.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        span.set_tag("error", true)
+        span.log_kv(key: "message", value: event.message)
```
Comment (mpetazzoni) [inline lib/mongodb/instrumentation/command_subscriber.rb:52]: Use the `error.msg` event key

**Review by mpetazzoni — CHANGES_REQUESTED** [review]

File: lib/mongodb/instrumentation/command_subscriber.rb (hunk 5 above)
Comment (achandras) [reply lib/mongodb/instrumentation/command_subscriber.rb:52]: Is it OK to deviate from the opentracing tags spec here? I think they just have the `message` tag for errors.

Comment (mpetazzoni) [reply lib/mongodb/instrumentation/command_subscriber.rb:52]: You're right, looks like they do have a spec around this: https://github.com/opentracing/specification/blob/master/semantic_conventions.md#captured-errors

Keep as is then

**Review by mpetazzoni — APPROVED** [review]

**Pull request merged**